third_party/3/pyspark/sql/functions.pyi

```diff
@@ -221,18 +221,18 @@ def variance(col: ColumnOrName) -> Column: ...
 
 class UserDefinedFunction:
     func = ...  # type: Callable[..., Any]
-    returnType = ...  # type: DataType
-    def __init__(self, func: Callable[..., Any], returnType: DataType, name: Optional[str] = ...) -> None: ...
+    returnType = ...  # type: DataTypeOrString
+    def __init__(self, func: Callable[..., Any], returnType: DataTypeOrString, name: Optional[str] = ...) -> None: ...
     def __call__(self, *cols: ColumnOrName) -> Column: ...
 
-def udf(f: Callable[..., Any], returnType: DataType = ...) -> Callable[..., Column]: ...
+def udf(f: Callable[..., Any], returnType: DataTypeOrString = ...) -> Callable[..., Column]: ...
 
 class PandasUDFType:
     SCALAR: int = ...
     GROUPED_MAP: int = ...
     GROUPED_AGG: int = ...
 
 @overload
-def pandas_udf(f: Callable[..., pandas.core.series.Series], returnType: StructField, functionType: int) -> Callable[..., Column]: ...
+def pandas_udf(f: Callable[..., pandas.core.series.Series], returnType: DataTypeOrString, functionType: int = ...) -> Callable[..., Column]: ...
```

Comment:
There are subtle interactions between optional `functionType` and `returnType` and I believe that these should be properly reflected with overloading. Additionally if you want to allow `functionType` to be optional, so should be `returnType`.

Reply:
I changed it now so that the variant that takes a (Pandas) Dataframe as a input only accepts `StructType`, I think that's the main influence of the `functionType` to `returnType`. Once we can use typing literals in Python3.8, we can express things more clearly here.

Regarding making `returnType` optional, I'd disagree, since that [fails](https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L3177-L3178). Whereas when you don't pass a `functionType`, it defaults to [`SCALAR`](https://github.com/apache/spark/blob/master/python/pyspark/sql/functions.py#L3175).

Reply:
> Regarding making returnType optional, I'd disagree, since that fails. Whereas when you don't pass a functionType, it defaults to SCALAR.

I stand corrected.